Python/specialize.c

```diff
@@ -778,6 +793,156 @@ _Py_Specialize_StoreAttr(PyObject *owner, _Py_CODEUNIT *instr, PyObject *name, S
+int
+_Py_Specialize_LoadMethod(PyObject *owner, _Py_CODEUNIT *instr, PyObject *name, SpecializedCacheEntry *cache)
+{
+    _PyAdaptiveEntry *cache0 = &cache->adaptive;
+    _PyAttrCache *cache1 = &cache[-1].attr;
+    _PyObjectCache *cache2 = &cache[-2].obj;
+
+    PyTypeObject *owner_cls = Py_TYPE(owner);
+    if (PyModule_CheckExact(owner)) {
+        int err = specialize_module_load_attr(owner, instr, name, cache0, cache1,
+            LOAD_METHOD, LOAD_METHOD_MODULE);
+        if (err) {
+            goto fail;
+        }
+        goto success;
+    }
+    if (owner_cls->tp_dict == NULL) {
+        if (PyType_Ready(owner_cls) < 0) {
+            return -1;
+        }
+    }
+    // Technically this is fine for bound method calls, but slightly slower at runtime.
+    if (Py_TYPE(owner_cls)->tp_dictoffset < 0) {
+        SPECIALIZATION_FAIL(LOAD_METHOD, SPEC_FAIL_NEGATIVE_DICTOFFSET);
+        goto fail;
+    }
+    PyObject **owner_dictptr = _PyObject_GetDictPtr(owner);
+    int owner_has_dict = (owner_dictptr != NULL && *owner_dictptr != NULL);
+    PyDictObject *owner_dict = owner_has_dict ? (PyDictObject *)*owner_dictptr : NULL;
+    // Check for classmethods.
+    int owner_is_class = (owner_cls->tp_getattro != PyObject_GenericGetAttr) &&
+        PyObject_TypeCheck(owner, &PyBaseObject_Type) && PyType_Check(owner);
+
+    PyObject *descr = NULL;
+    DesciptorClassification kind = 0;
+    if (owner_is_class) {
+        // class method
+        kind = analyze_descriptor((PyTypeObject *)owner, name, &descr, 0);
+        // Store the version right away, in case it's modified halfway through.
+        cache1->tp_version = ((PyTypeObject *)owner)->tp_version_tag;
+    }
+    else {
+        // instance method
+        kind = analyze_descriptor(owner_cls, name, &descr, 0);
+        cache1->tp_version = owner_cls->tp_version_tag;
+    }
+
+    if (kind != METHOD || descr == NULL) {
+#if SPECIALIZATION_STATS
+        if (owner_is_class) {
+            if (descr == NULL) {
+                SPECIALIZATION_FAIL(LOAD_METHOD, SPEC_FAIL_NOT_METHOD);
+                goto fail;
+            }
+            // Builtin METH_CLASS class method -- ie wrapped PyCFunction
+            else if (kind == NON_OVERRIDING &&
+                Py_IS_TYPE(descr, &PyClassMethodDescr_Type)) {
+                // NOTE (KJ): Don't bother, rare and no speedup in microbenchmarks.
+                SPECIALIZATION_FAIL(LOAD_METHOD, SPEC_FAIL_BUILTIN_CLASS_METHOD);
+                goto fail;
+            }
+            else if (Py_IS_TYPE(descr, &PyClassMethod_Type)) {
+                // Note: this is the actual Python classmethod(func) object.
+                SPECIALIZATION_FAIL(LOAD_METHOD, SPEC_FAIL_CLASS_METHOD_OBJ);
+                goto fail;
+            }
+        }
+#endif
+        SPECIALIZATION_FAIL(LOAD_METHOD, SPEC_FAIL_NOT_METHOD);
+        goto fail;
+    }
+    PyObject **cls_dictptr = _PyObject_GetDictPtr((PyObject *)owner_cls);
```

Comment:
Why is this necessary? Doesn't `analyze_descriptor` do this for us?

Reply:
Correct. I forgot that `kind` would become something else. Thanks.